actionpack/lib/action_dispatch/http/uri.rb

```diff
@@ -0,0 +1,247 @@
+# frozen_string_literal: true
+
+require "active_support/core_ext/module/attribute_accessors"
+require "uri/generic"
+
+class ActionDispatch::Http::URI
+  IP6_REGEX       = /(https?)?[:\/]*([\[]?(?:[0-9a-fA-F]{1,4}:){7}[0-9a-fA-F]{1,4}[\]]?)[:]?([0-9]*)(.*)/
+  IP_HOST_REGEXP  = /\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}$/
+  HOST_REGEXP     = /(^[^:]+:\/\/)?(\[[^\]]+\]|[^:]+)(?::(\d+$))?/
+  PROTOCOL_REGEXP = /^([^:]+)(:)?(\/\/)?$/
+
+  mattr_accessor :secure_protocol, default: false
+  mattr_accessor :tld_length, default: 1
+
+  delegate :scheme, :host, :port, :path, :query, :fragment, :to_s, to: :uri
+  delegate :scheme=, :host=, :port=, :path=, :query=, :fragment=, to: :uri
```

Comment:
Absent a compelling use case, I think we should probably make this class immutable, and drop these `=` methods.

Reply:
in the long term it would be nice to have a URI object for the current request, this PR is a first step, but more should follow.

Reply:
Yes, I agree that's a sensible goal.. but I don't think the current request URI needs to be mutable.

Reply:
Yes, we can re add it in the future if needed.